src/custom_executor.rs

```diff
@@ -0,0 +1,71 @@
+use std::{any::Any, future::Future, pin::Pin};
+
+use crate::{error::Error, ComputeHeavyFutureExecutor};
+
+/// The input future for a custom executor closure.
+/// This is the regular future, with its output type erased to Any.
+pub type AnyWrappedFuture = Pin<Box<dyn Future<Output = Box<dyn Any + Send + 'static>> + Send + 'static>>;
+
+/// A closure that accepts a type-erased input future and returns a future that resolves to
+/// either Ok(<type erased input future's output>) or Err(<boxed error representing executor errors>)
+pub type CustomExecutorClosure = Box<
+    dyn Fn(
+            AnyWrappedFuture,
+        ) -> Box<
+            dyn Future<Output = Result<Box<dyn Any + Send + 'static>, Box<dyn std::error::Error + Send + Sync>>>,
+        > + Send
+        + Sync,
+>;
+
+// used for checking that the closure doesn't mutate the output type while working with Any
+struct PrivateType {}
+
+pub(crate) struct CustomExecutor {
+    pub(crate) closure: CustomExecutorClosure,
+}
+
+impl CustomExecutor {
+    pub(crate) async fn new(closure: CustomExecutorClosure) -> Self {
+        let executor = Self { closure };
+
+        let test_future = async { PrivateType {} };
+
+        if let Err(_) = executor.execute(test_future).await {
+            panic!(
+                "CustomExecutor strategy initialized with a closure that \
+            changes the future's output type"
+            );
+        }
+
+        executor
+    }
+}
+
+impl ComputeHeavyFutureExecutor for CustomExecutor {
+    async fn execute<F, O>(&self, fut: F) -> Result<O, Error>
+    where
+        F: Future<Output = O> + Send + 'static,
+        O: Send + 'static,
+    {
+        let wrapped_future = Box::pin(async move {
+            let res = fut.await;
+            Box::new(res) as Box<dyn Any + Send>
```

Comment:
Not sure whether using `Any` or a channel is smarter here.

Reply:
`Any` would no longer be agnostic to underlying caller's custom behavior right?

Reply:
We discussed, the channel would let us erase the output type and replace with `()` at the cost of slightly more overhead from the channel. But then the input API would be nicer probably since less boxing by the caller.

Will poke around with this.

Reply:
I have it working fine with the channel, but I was thinking more, and one use case I want to support is selective handling based on future type (for prioritization, etc). So totally type erasing the future's output type might not be ideal. Will poke around to see if there is a middle ground that preserves ergonomics.

Reply:
Ehh, I decided not to scope creep on this. We can always add another strategy or a toggle on the existing one if this becomes an important use case.